pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -260,6 +274,19 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
         _jsonIndexCreatorMap.put(columnName, jsonIndexCreator);
       }
 
+      H3IndexConfig h3IndexConfig = h3IndexConfigs.get(columnName);
+      if (h3IndexConfig != null) {
+        Preconditions
+            .checkState(fieldSpec.isSingleValueField(), "H3 index is currently only supported on single-value columns");
+        Preconditions.checkState(fieldSpec.getDataType() == DataType.BYTES,
+            "H3 index is currently only supported on BYTES columns");
+        H3IndexResolution resolution = h3IndexConfig.getResolution();
+        GeoSpatialIndexCreator h3IndexCreator =
+            segmentCreationSpec.isOnHeap() ? new OnHeapH3IndexCreator(_indexDir, columnName, resolution)
```

Comment:
any scenario to use on heap index?

Reply:
When memory is not an issue (e.g. hadoop), on-heap creator is more efficient